source/MaterialXTest/ShaderValid.cpp

```diff
@@ -466,9 +508,450 @@ TEST_CASE("GLSL geometry", "[shadervalid]")
     shaderGenerator->registerSourceCodeSearchPath(searchPath);
 
     // Run validation
-    runValidation(nodeGraph->getName(), outputList, validator, shaderGenerator, orthographicView, doc, log);
+    runValidation("geometry_attributes", outputList, validator, shaderGenerator, orthographicView, docs, log);
 }
 
+TEST_CASE("GLSL math", "[shadervalid]")
+{
+  #ifdef LOG_TO_FILE
+      std::ofstream logfile("log_shadervalid_glsl_geometry.txt");
+      std::ostream& log(logfile);
+  #else
+      std::ostream& log(std::cout);
+  #endif
+
+      mx::FilePath searchPath = mx::FilePath::getCurrentPath() / mx::FilePath("documents/Libraries");
+
+      // Stores our documents
+      std::vector<mx::DocumentPtr> docs;
+
+      // Store graph outputs that test some math operator nodes
+      std::vector<mx::ElementPtr> outputList;
```

Comment:
All the test graphs below are created programatically but I'm not sure I see the reason for this. It's a lot of code for something what would be easier and more flexible to do in our editor (BlackSky): create the node graphs there, save out a document and make this test document part of a test suite. Your test here could then just load each document in the test suite and generate shaders for all node graphs found. It's then data driven and extendable without code changes and rebuilds.

It's another thing if the node graphs created here were generated from the library nodedefs, so that all nodes and type variants where created and tested automagically. But now it seems to be just hard coded node creation, which feels a lot easier to do in an editor.

Reply:
Hey Niklas, I thought so too, but Bernard seemed to think this way was better. I’ll let him speak for himself, but I think the rationale was that this way if the MaterialX Spec changes you’re not just left with a bunch of broken MaterialX documents. It definitely seemed like a bit more work this way to me.

Reply:
Just as a side note, it’s worth mentioning that MaterialX documents are intended to be forward-compatible, even as the specification evolves over time.  Our internal material library contains documents from a range of different versions (some as old as v1.23), and they’re promoted to the latest version at import time.

Reply:
Exactly, specification changes are handled nicely by the document upgrade feature at file load time, which is another good reason for keeping these tests data driven. But let's discuss this more offline.